src/NServiceBus.Core/Routing/ApplyReplyToAddressBehavior.cs

```diff
@@ -2,23 +2,96 @@
 {
     using System;
     using System.Threading.Tasks;
+    using NServiceBus.Extensibility;
     using NServiceBus.Pipeline;
 
     class ApplyReplyToAddressBehavior : Behavior<IOutgoingLogicalMessageContext>
     {
-        public ApplyReplyToAddressBehavior(string replyToAddress)
+        public ApplyReplyToAddressBehavior(string sharedQueue, string instanceSpecificQueue, string publicReturnAddress, string distributorAddress)
         {
-            this.replyToAddress = replyToAddress;
+            this.sharedQueue = sharedQueue;
+            this.instanceSpecificQueue = instanceSpecificQueue;
+            this.publicReturnAddress = publicReturnAddress;
+            this.distributorAddress = distributorAddress;
         }
 
         public override Task Invoke(IOutgoingLogicalMessageContext context, Func<Task> next)
         {
-            context.Headers[Headers.ReplyToAddress] = replyToAddress;
+            var state = context.Extensions.GetOrCreate<State>();            
+            if (state.Option == RouteOption.RouteToThisInstance && instanceSpecificQueue == null)
+            {
+                throw new InvalidOperationException("Cannot route a reply to this specific instance because endpoint instance ID was not provided by either host, a plugin or user. You can specify it via BusConfiguration.EndpointInstanceId, use a specific host or plugin.");
+            }
+            context.Headers[Headers.ReplyToAddress] = ApplyUserOverride(publicReturnAddress ?? GetDefaultReplyToValue(context), state);
 
+            //Legacy distributor logic
+            if (state.FromDistributor)
+            {
+                context.Headers[Headers.ReplyToAddress] = distributorAddress;
+            }
             return next();
         }
+        
 
-        string replyToAddress;
+        string ApplyUserOverride(string replyTo, State state)
+        {
+            if (state.Option == RouteOption.RouteToAnyInstanceOfThisEndpoint)
+            {
+                replyTo = sharedQueue;
+            }
+            else if (state.Option == RouteOption.RouteToThisInstance)
+            {
+                replyTo = instanceSpecificQueue;
+            }
+            else if (state.Option == RouteOption.ExplicitDestination)
+            {
+                replyTo = state.ExplicitDestination;
+            }
+            return replyTo;
+        }
 
+        string GetDefaultReplyToValue(IExtendable context)
+        {
+            string receivedFrom;
+            return context.Extensions.TryGet("ReceivedFrom", out receivedFrom) 
+                ? receivedFrom 
```

Comment:
My reasoning was if somebody sent you a message on your specific queue (for any reason), he probably new only you can handle it so the whole "conversation" should stick to that specific instance. Feel free to challenge that reasoning.